src/java/org/apache/cassandra/db/compaction/UnifiedCompactionStrategy.java

```diff
@@ -129,35 +182,61 @@ public Collection<Collection<CompactionSSTable>> groupSSTablesForAntiCompaction(
-        // space to complete.
-        // The reason for this is to ensure that shard-spanning sstables get compacted with everything they overlap with.
-        // For example, if an sstable on L0 covers shards 0 to 3, a sharded maximal compaction will compact that with
-        // higher-level sstables in shard 0 and will leave a low-level sstable containing the non-shard-0 data from that
-        // L0 sstable. This is a non-expected and non-desirable outcome.
+        // The tasks are split by repair status and disk, as well as in non-overlapping sections to enable some
+        // parallelism (to the amount that L0 sstables are split, i.e. at least base_shard_count). The result will be
+        // split across shards according to its density. Depending on the parallelism, the operation may require up to
+        // 100% extra space to complete.
         List<AbstractCompactionTask> tasks = new ArrayList<>();
-        for (Shard shard : getCompactionShards(realm.getLiveSSTables(), arenaSelector.nonShardingSelector(), true))
+        for (Arena arena : getCompactionArenas(realm.getLiveSSTables(), true))
         {
-            LifecycleTransaction txn = realm.tryModify(shard.sstables, OperationType.COMPACTION);
-            if (txn != null)
-                tasks.add(createCompactionTask(gcBefore, txn, true, splitOutput));
+            List<Set<CompactionSSTable>> nonOverlapping = splitInNonOverlappingSets(arena.sstables);
+            for (Set<CompactionSSTable> set : nonOverlapping)
+            {
+                LifecycleTransaction txn = realm.tryModify(set, OperationType.COMPACTION);
+                if (txn != null)
+                    tasks.add(createCompactionTask(gcBefore, txn, true, splitOutput));
+            }
         }
         return CompactionTasks.create(tasks);
     }
 
+    private static List<Set<CompactionSSTable>> splitInNonOverlappingSets(List<CompactionSSTable> sstables)
+    {
+        List<Set<CompactionSSTable>> overlapSets = constructOverlapSets(sstables,
+                                                                        UnifiedCompactionStrategy::startsAfter,
+                                                                        CompactionSSTable.firstKeyComparator,
+                                                                        CompactionSSTable.lastKeyComparator);
+        Set<CompactionSSTable> group = overlapSets.get(0);
+        List<Set<CompactionSSTable>> groups = new ArrayList<>();
+        for (int i = 1; i < overlapSets.size(); ++i)
+        {
+            Set<CompactionSSTable> current = overlapSets.get(i);
+            if (Sets.intersection(current, group).isEmpty())
+            {
+                groups.add(group);
+                group = current;
+            }
+            else
+            {
+                group.addAll(current);
+            }
+        }
+        groups.add(group);
+        return groups;
```

Comment:
These are sets of sstables that can be compacted in parallel, correct?

Reply:
Yes. Since they don't overlap, there can't be overwrites or deletions that affect one group but reside in another.

If the split points change (e.g. because of changes in local ownership), the result of one single all-encompassing compaction may be different from independent compaction of these groups, but only in the splitting points &ndash; as the latter will keep the existing ones.